src/main/java/com/google/sps/servlets/DashboardServlet.java

```diff
@@ -14,62 +14,125 @@
+    
+    // combine the two filters with the 'and' operator 
+    final Query timedQuery = new Query(DBUtilities.INTERACTION_TABLE).setFilter(CompositeFilterOperator.and(startTimestampFilter, endTimestampFilter));
+    PreparedQuery interactions = datastore.prepare(timedQuery);
 
     Field[] allFields = Property.class.getDeclaredFields();
+    Property classInstance = new Property();    
   
-    HashMap<String, Double> interactionPercentages = new HashMap<>();
+    HashMap<String, Double> dataToSend = new HashMap<>();
 
-    for (Field field : allFields) {
-      String property = (String) field.get(new Property()); // gets the value of the field variable
+    // start iterating at 2 to skip timestamp and correlator properties
+    for (int i = 2; i < allFields.length; ++i){
+      String property = (String) allFields[i].get(classInstance); // gets the value of the field variable
+
+      Filter propertyFilter =  new FilterPredicate(property, FilterOperator.EQUAL, true);
+      Query filteredQuery = new Query(DBUtilities.INTERACTION_TABLE);
+      filteredQuery.setFilter(CompositeFilterOperator.and(propertyFilter, startTimestampFilter, endTimestampFilter));
 
-      Filter keyFilter =  new FilterPredicate(property, FilterOperator.EQUAL, true);
-      Query filteredQuery = new Query(DBUtilities.INTERACTION_TABLE).setFilter(keyFilter);
-    
       int numUsersInteracted = datastore.prepare(filteredQuery).countEntities();
       int totalInteractions = interactions.countEntities();
 
-      interactionPercentages.put(property, numUsersInteracted / (double) totalInteractions);
+      double percentage = totalInteractions == 0 ? 0 : numUsersInteracted / (double) totalInteractions;
+
+      dataToSend.put(property, percentage);
     }
 
-    return interactionPercentages;
+    dataToSend.put("totalInteractions", (double) interactions.countEntities());
+    
+    return dataToSend;
   }
 
+ /**
+  * Given the data to convert to json, this function returns a json
+  * representation.
+  *
+  * @param data A hashmap from a string to a double to be converted to a json format.
+  * @return A json representation of the input data.
+  */
   private String convertToJson(HashMap<String, Double> data) {
     Gson gson = new Gson();
     String json = gson.toJson(data);
     return json;
   }
+
+ /**
+  * Given the start and end timestamps requested, this function validates and sets 
+  * the instance variables.
+  *
+  * @param startTimestamp A string containing the requested start time as a UNIX timestamp
+  * @param endTimestamp A string containing the requested end time as a UNIX timestamp
+  */
+  private void validateAndSetVariables(String requestStartTimestamp, String requestEndTimestamp){
+
+    if (requestStartTimestamp.isEmpty() || requestEndTimestamp.isEmpty()) {
+      this.startTimestamp = DEFAULT_START_TIMESTAMP;
```

Comment:
There shouldn't be any need to use class variables in something like this.  See if you can refactor it so that you don't need to save anything to the class.  This should be able to process many simultaneous requests. I don't know how using class variables would effect that, but probably not well

Reply:
removed class instance variables